bpfd/src/bpf.rs

```diff
@@ -207,148 +135,205 @@ impl<'a> BpfManager<'a> {
         Ok(())
     }
 
-    pub(crate) fn add_program(
-        &mut self,
-        iface: String,
-        path: String,
-        priority: i32,
-        section_name: String,
-        proceed_on: Vec<i32>,
-        owner: String,
-    ) -> Result<Uuid, BpfdError> {
-        let if_index = self.get_ifindex(&iface)?;
-        let id = Uuid::new_v4();
-        let map_pin_path = format!("/var/run/bpfd/fs/maps/{}", id);
-        fs::create_dir_all(map_pin_path.clone())?;
-
-        let mut ext_loader = BpfLoader::new()
-            .extension(&section_name)
-            .map_pin_path(map_pin_path.clone())
-            .load_file(&path)?;
-
-        ext_loader.program_mut(&section_name).ok_or_else(|| {
-            let _ = fs::remove_dir_all(map_pin_path);
-            BpfdError::SectionNameNotValid(section_name.clone())
-        })?;
-
-        // Calculate the next_available_id
-        let next_available_id = if let Some(prog) = self.programs.get(&if_index) {
-            prog.len() + 1
-        } else {
-            self.programs.insert(if_index, HashMap::new());
-            1
-        };
-        if next_available_id > 10 {
-            return Err(BpfdError::TooManyPrograms);
+    pub(crate) fn add_program(&mut self, p: Program) -> Result<Uuid, BpfdError> {
+        match p {
+            Program::Xdp(_, _) => self.add_xdp_program(p),
+            Program::Tracepoint(_, _) => self.add_single_attach_program(p),
         }
+    }
 
-        // Calculate the dispatcher revision
-        let (old_revision, revision) = if let Some(old_revision) = self.revisions.remove(&if_index)
-        {
-            let r = old_revision.wrapping_add(1);
-            self.revisions.insert(if_index, r);
-            (Some(old_revision), r)
-        } else {
-            self.revisions.insert(if_index, 0);
-            (None, 0)
-        };
+    pub(crate) fn add_xdp_program(&mut self, p: Program) -> Result<Uuid, BpfdError> {
+        if let Program::Xdp(ref data, ref info) = p {
+            let if_index = info.if_index;
+            let id = Uuid::new_v4();
+            let map_pin_path = format!("/var/run/bpfd/fs/maps/{}", id);
+            fs::create_dir_all(map_pin_path.clone())?;
+
+            let mut ext_loader = BpfLoader::new()
+                .extension(&data.section_name)
+                .map_pin_path(map_pin_path.clone())
+                .load_file(&data.path)?;
+
+            ext_loader.program_mut(&data.section_name).ok_or_else(|| {
+                let _ = fs::remove_dir_all(map_pin_path);
+                BpfdError::SectionNameNotValid(data.section_name.clone())
+            })?;
+
+            // Calculate the next_available_id
+            let next_available_id = self
```

Comment:
To make sure I'm understanding this.  It looks like you are calculating the number of programs of a given type on a given interface (in this case) by iterating through the entire list of programs and selecting the ones of the type and interface you're looking for.  Is that correct?

Reply:
This has been refactored slightly in the latest patchset but the concept remains the same...
We walk the entire programs list twice - filtering only the programs we want, like XDP for if_index 3.
First pass is to sort the programs and update the `current_position`, second time is to create a copy which gets passed to `Dispatcher::new()`.
That can probably be collapsed into a single iteration but my Rust skills are failing me there.